prototypes/ScheduledMerges.hs

```diff
@@ -972,48 +986,79 @@ lookupsTree k = go
     lookupBatch' = lookupBatch Nothing k
 
 -------------------------------------------------------------------------------
--- Updates
+-- Nominal credits
 --
 
+newtype NominalCredit = NominalCredit Credit
+  deriving stock Show
+
+newtype NominalDebt = NominalDebt Credit
+  deriving stock Show
+
 -- TODO: If there is a UnionLevel, there is no (more expensive) last level merge
 -- in the regular levels, so a little less merging work is required than if
 -- there was no UnionLevel. It might be a good idea to spend this "saved" work
 -- on the UnionLevel instead. This makes future lookups cheaper and ensures that
 -- we can get rid of the UnionLevel at some point, even if a user just keeps
 -- inserting without calling 'supplyUnionCredits'.
-supplyCreditsLevels :: Credit -> Levels s -> ST s ()
-supplyCreditsLevels unscaled =
+supplyCreditsLevels :: NominalCredit -> Levels s -> ST s ()
+supplyCreditsLevels nominalDeposit =
   traverse_ $ \(Level ir _rs) -> do
     case ir of
       Single{} -> return ()
-      Merging mp mr -> do
-        factor <- creditsForMerge mp mr
-        let credits = ceiling (fromIntegral unscaled * factor)
-        when (credits > 0) $ do
-          _ <- supplyCreditsMergingRun credits mr
-          -- we don't mind leftover credits, each level completes independently
-          return ()
-
--- | The general case (and thus worst case) of how many merge credits we need
--- for a level. This is based on the merging policy at the level.
---
-creditsForMerge :: MergePolicy -> MergingRun t s -> ST s Rational
+      Merging _mp nominalDebt nominalCreditVar
+              mr@(MergingRun _  physicalDebt _) -> do
+
+        nominalCredit       <- depositNominalCredit
+                                 nominalDebt nominalCreditVar nominalDeposit
+        physicalCredit      <- suppliedCreditMergingRun mr
+        let !physicalCredit' = scaleNominalToPhysicalCredit
+                                 nominalDebt physicalDebt nominalCredit
+            -- Our target physicalCredit' could actually be less than the
+            -- actual current physicalCredit if other tables were contributing
+            -- credits to the shared merge.
+            !physicalDeposit = physicalCredit' - physicalCredit
+
+        -- So we may have a zero or negative deposit, which we ignore.
+        when (physicalDeposit > 0) $ do
+          leftoverCredits <- supplyCreditsMergingRun physicalDeposit mr
+          -- For merges at ordinary levels (not unions) we expect to hit the
+          -- debt limit exactly and never exceed it.
+          assert (leftoverCredits == 0) $ return ()
```

Comment:
Do you have an idea of whether it would be problematic for async exceptions to occur between the nominal credit deposit and supplying the credits to the merging run?

Reply:
I've added a big comment. It's a design trade-off. There's a race but I think quite a benign one, or we have to support supplying merge credit in either relative or absolute styles.